internal/pkg/peer/blocksprovider/bft_deliverer.go

```diff
@@ -85,76 +123,123 @@ func (d *BFTDeliverer) DeliverBlocks() {
+			break FetchAndMonitorLoop
+		}
 		// start a block fetcher and a monitor
 		// a buffered channel so that the fetcher goroutine can send an error and exit w/o waiting for it to be consumed.
 		d.fetchErrorsC = make(chan error, 1)
 		source := d.fetchSources[d.fetchSourceIndex]
 		go d.FetchBlocks(source)
 
-		// TODO start a censorship monitor
+		// create and start a censorship monitor
+		d.censorshipMonitor = d.CensorshipDetectorFactory.Create(
+			d.ChannelID, d.BlockVerifier, d.requester, d, d.fetchSources, d.fetchSourceIndex, timeoutConfig)
+		go d.censorshipMonitor.Monitor()
 
 		// wait for fetch  errors, censorship suspicions, or a stop signal.
 		select {
 		case <-d.DoneC:
 			break FetchAndMonitorLoop
 		case errFetch := <-d.fetchErrorsC:
-			if errFetch != nil {
-				switch errFetch.(type) {
-				case *errStopping:
-					// nothing to do
-				case *errRefreshEndpoint:
-					// get new endpoints and reassign fetcher and monitor
-					d.refreshSources()
-					d.resetFetchFailureCounter()
-				default:
-					d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
-					d.incFetchFailureCounter()
-				}
-				// TODO can it be nil?
+			d.Logger.Debugf("Error received from fetchErrorsC channel: %s", errFetch)
+
+			switch errFetch.(type) {
+			case *ErrStopping:
+				// nothing to do
+				break FetchAndMonitorLoop
+			case *errRefreshEndpoint:
+				// get new endpoints and reassign fetcher and monitor
+				d.refreshSources()
+				d.resetFetchFailureCounter()
+			case *ErrFatal:
+				d.Logger.Errorf("Failure in FetchBlocks, something is critically wrong: %s", errFetch)
+				break FetchAndMonitorLoop
+			default:
+				d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
+				d.incFetchFailureCounter()
 			}
-		case errMonitor := <-d.monitorErrorsC:
-			// TODO until we implement the censorship monitor this nil channel is blocked
-			if errMonitor != nil {
-				d.Logger.Warningf("Censorship suspicion: %s", err)
-				// TODO close the block receiver, increment the index
-				// TODO
+		case errMonitor := <-d.censorshipMonitor.ErrorsChannel():
+			d.Logger.Debugf("Error received from censorshipMonitor.ErrorsChannel: %s", errMonitor)
+
+			switch errMonitor.(type) {
+			case *ErrStopping:
+				// nothing to do
+				break FetchAndMonitorLoop
+			case *ErrCensorship:
+				d.Logger.Warningf("Censorship suspicion: %s", errMonitor)
+				d.mutex.Lock()
+				d.blockReceiver.Stop()
+				d.mutex.Unlock()
+				d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
+				d.incFetchFailureCounter()
+			case *ErrFatal:
+				d.Logger.Errorf("Failure in CensorshipMonitor, something is critically wrong: %s", errMonitor)
+				break FetchAndMonitorLoop
+			default:
+				d.Logger.Errorf("Unexpected error from CensorshipMonitor, something is critically wrong: %s", errMonitor)
+				break FetchAndMonitorLoop
 			}
-			// TODO can it be nil?
 		}
+
+		d.censorshipMonitor.Stop()
```

Comment:
why do we start everything just to tear it down? Can't we just stop one censorship monitor?

Reply:
the `d.censorshipMonitor.Stop()` at the end of the loop is only called when something "bad" happens, that is: 

- block fetch error
- censorship event

These are hopefully rare events. Therefore, for simplicity, it is easier to close all the header receivers and and block receiver reassign everything from scratch than to complicate the implementation and 
- close only the block receiver and replace it with a header receive
- close a header receiver and replace it with a block receiver

Note that the censorship monitor is a collection of header receivers, yet failures in header receivers do not cause a restart of the monitor. Header receiver failures are handled within. This is because eventually, an incorrect orderer (malicious) will be assigned to a header receiver, which should be able to tolerate its behavior and not disrupt the reception of blocks.